### PR TITLE
Add match phrase prefix query

### PR DESCRIPTION
**Pull request by arielshad**

### CODE REVIEW

**Comment by arielshad** [comment]
there is one issue, the build ran on windows machine and caused problems with the slashes in the path of the import statment.. need to figure it out..

**Comment by danpaz** [comment]
Thanks for the PR @arielshad! Mind adding a description here with the changes proposed? Looks like a few things got added that I didn't follow e.g. what's the use case for `customQuery`?

**Comment by arielshad** [comment]
Hey @danpaz ,
after I wrote the [match-phrase-prefix](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-match-query.html#query-dsl-match-query-phrase-prefix) query, 
I came in a need for a [match_phrase](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-match-query.html#query-dsl-match-query-phrase) query, 
I saw that the structure and syntax are pretty much the same and there more query like that, that I might want to use in the future, instead of writing another wrapper I decied to create a customQuery which will get the queryName and build the query you can see example in the test case [here](https://github.com/arielshad/bodybuilder/blob/Sentic_-_bodybuilder.ks/test/queries/custom-query.js)

does that make any sense? :+1:  :-1:

File: package.json [hunk 1]
```diff
@@ -21,18 +21,18 @@
...
+    "babel-preset-es2015": "^6.13.2",
     "chai": "3.2.0",
     "documentation": "3.0.4",
     "eslint": "1.10.2",
     "mocha": "2.4.5",
     "webpack": "1.12.13"
   },
   "dependencies": {
-    "lodash": "4.9.0"
+    "lodash": "^4.9.0"
```
Comment (danpaz) [inline package.json]: Can you revert these updates in package.json? I think it's really important to pin dependencies especially when working with fast moving projects like babel, and upgrades to this kind of tooling should go in a separate PR, if they are required.

File: src/aggregations/filter-aggregation.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * @return {Object}       Script Aggregation.
+ */
+export default function filterAggregation(filterName, filterRef, operator, value) {
+  return {
+    [filterName]: {
+      bucket_selector: {
+        buckets_path: {
+          [filterName]: filterRef
+        },
+        script: `${filterName} ${operator} ${value}`
```
Comment (danpaz) [inline src/aggregations/filter-aggregation.js]: This is great, but is focused really heavily on using this aggregation with an inline script and bucket_selector, which is not always the desired approach https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-filter-aggregation.html.

I think we can support these more advanced options, but it should be an opt-in custom option. So I'm thinking a function signature like

```
function filterAggregation(filterType, filterField, filterValue)
```

would be ideal here. What do you think?

File: src/queries/match-phrase-prefix.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * Construct a Match Phrase Prefix query.
+ *
+ * @param  {String} field Field name to query over.
+ * @param  {String} term  Query value.
+ * @param  {Object} opts  See match_phrase_prefix
+ * @return {Object}       Match match_phrase_prefix.
+ */
+export default function matchPhrasePrefix(field, term, opts = {}) {
```
Comment (danpaz) [inline src/queries/match-phrase-prefix.js]: This looks good and would be happy to look at it in a separate PR to release it more quickly.

**Comment by danpaz** [comment]
Thanks for clarifying @arielshad, there's some really good stuff here but in general combining a grab-bag of features into one PR makes it harder to review and will take longer to approve and release. So a PR-per-feature approach, with a description of the change and what problem it solves, works way better. 

I left some comments, feel free to work off this branch or open a new one if that's easier. Hope this helps move things forward!

**Comment by cacois** [comment]
agreed on grab bag PRs. I was going to write a wildcard query, but see that your custom query would alleviate the need. If you separated these features into multiple PRs, I'm sure many would be merged immediately while discussion continued on some of the more complex features.

**Comment by danpaz** [comment]
Closing as bodybuilder 2 will support all of these features once out of beta. Check it out! https://github.com/danpaz/bodybuilder/releases/tag/v2.0.0-beta.1
